cmd/forwarder/fixer.go

```diff
@@ -21,48 +22,68 @@ const (
 	maxMsgidLength    = 32
 )
 
-var nilVal = []byte(`- `)
-var queryParams = []string{"index", "source", "sourcetype"}
-
 // Get metadata from the http request.
 // Returns an empty byte array if there isn't any.
-func getMetadata(req *http.Request, cred *credential, metadataId string) ([]byte, bool) {
+func getMetadata(req *http.Request, cred *credential, metadataId string, queryFieldParams []string) ([]byte, bool) {
 	var metadataWriter bytes.Buffer
 	var foundMetadata bool
+
 	// Calculate metadata query parameters
 	if metadataId != "" {
+		var queryParams = []string{"index", "source", "sourcetype"}
+		var fieldsBuilder strings.Builder
+
+		// Pre-grow to typical lengths to avoid extra slice allocations
+		fieldsBuilder.Grow(128)
+		metadataWriter.Grow(1024)
```

Comment:
Nit (or at least, for discusson: Don't consider this a blocker).

It might be good to have somebody who's better at `go` than I am (and I'm sure we have lots of those) take a look at this :). 

Why are these lengths "typical"? Given the agnostic nature of this code, that seems hard to be sure of (and I'm not sure that we've ever actually done analysis of the typical lengths of these strings that we see in our production environment).

Another thought is that it might be worth using a pool of stringBuilders here (or a pool of structs with purpose-specific elements): If we have a pool with a max size of the total number of goroutines, then we don't need to worry about preallocating a size: We just let the builders grow as big as they want, but we know that eventually as a builder gets used across multiple invocations, it will be as big as it will ever get.

☝️ I've never actually done this in `go`, but I think it's a thing people do.

Reply:
It was a just an easy efficiency thing so the underlying builder slice doesn't have to keep allocating more buffer space on every write.  It doesn't cost anything except a little extra memory.  Without looking that closely at all our logs I think we can safely assume they are going to be bigger than 0 at least.  Every time we leave this function we lose the builder and its underlying buffer so it's not going to keep it around at whatever size it was before.

Reply:
If it makes you feel better I can remove the "typical lengths" phrase :)